### PR TITLE
swap,asset,db,etc: swap state and clean swap shutdown

### PR DESCRIPTION
**Pull request by chappjc**
This addresses Swapper state persistence (save and load) and graceful Swapper shutdown.  This resends of ack requests for which responses were pending, relaunches of active coin waiters, and rest restores matches.

[TODO: more on `swap` changes here]

**Required changes to other packages:**

**`asset`**

The `asset` backends needed support for general outputs, not just UTXOs, and for possibly spent swap contracts.  New `Output` types in each asset backend are created as a base for most other types based on outputs, including `UTXO` and `Contract`. This changes many of the `server/asset` interfaces:
- Add to `Backend` the `Input`, `Output`, and `UnspentContract` interfaces.
- Move `Value` and `FeeRate` into the `Coin` interface.
- Create the `Output` interface, adding the `Addresses` method to `Coin`, constructed by `Backend.Output`.
- Rename `Contract` interface methods, and add `RefundAddress`.

Also, in `server/asset/dcr`, experiment with a more efficient `blockChans` structure using a map so that these channels can be easily removed when consumers go away (see `(dcr *Backend).Run`).

**`db`**

Add `Fatal() <-chan struct{}` method to `DEXArchivist`, which provides select semantics similar to `Context.Done` when there is a fatal backend error. Use `LastErr` to get the error.

**`market`**: Just a fix to `collectPreimages` where a down connection will never register a miss or a collected preimage.

**`auth,comms`**: Similar to the `market` fix, on send/request failure remove response handler, avoiding the wait for expiration, and make the caller handle the error, which may or may not be similar to the expire function.

### CODE REVIEW

File: server/market/market.go [hunk 1]
```diff
@@ -1051,15 +1051,20 @@ func (m *Market) collectPreimages(orders []order.Order) (cSum []byte, ordersReve
...
+			// since it will block if run twice.
+			missOnce.Do(func() { piChan <- nil })
 		}
 
 		// Send the preimage request to the order's owner.
 		err = m.auth.RequestWithTimeout(ord.User(), req, func(_ comms.Link, msg *msgjson.Message) {
 			m.handlePreimageResp(msg, reqData)
 		}, piTimeout, miss)
 		if err != nil {
+			miss() // only called by RequestWithTimeout on expire, not send errors
```
Comment (chappjc) [inline server/market/market.go:1080]: Was a bug where a request error (not just expire) would prevent a result for this requests, preventing epoch processing.

File: dex/ws/wslink.go [hunk 2]
```diff
@@ -70,28 +70,51 @@ type WSLink struct {
...
+func (c *WSLink) SendNow(msg *msgjson.Message) error {
+	writeErrChan := make(chan error, 1)
+	if err := c.send(msg, writeErrChan); err != nil {
+		return err
+	}
+	return <-writeErrChan
+}
+
+// CONCEPT
+func (c *WSLink) send(msg *msgjson.Message, writeErr chan<- error) error {
```
Comment (chappjc) [inline dex/ws/wslink.go:117]: There are paths where blocking is acceptable (the client's own websocket handler goroutines) and paths where it is not (requests from swapper, etc.).
I'm not using `SendNow`, but it's an idea that might improve things that presently have to wait for expire even though a write error is hit immediately.

File: server/asset/common.go [hunk 3]
```diff
@@ -75,21 +90,23 @@ type FundingCoin interface {
...
 	SpendSize() uint32
-	// Value is the output value.
-	Value() uint64
+}
+
+// Output is a transaction output.
+type Output interface {
+	Coin
+	Addresses() []string
 }
```
Comment (chappjc) [inline server/asset/common.go:99]: We discussed eliminating this interface, just using Coin for coins that represent tx outputs.

File: dex/msgjson/types.go [hunk 4]
```diff
@@ -365,7 +366,7 @@ func (msg *Message) String() string {
 
 // Match is the params for a DEX-originating MatchRoute request.
 type Match struct {
-	signable
+	Signature
```
Comment (chappjc) [inline dex/msgjson/types.go:369]: Otherwise can't gob encode/decode even though the `Sig` field was exported.

File: server/comms/comms_test.go [hunk 5]
```diff
@@ -329,28 +329,35 @@ func TestClientRequests(t *testing.T) {
 		if err != nil {
 			t.Fatalf("ban route send error: %v", err)
 		}
+		// Banish does Disconnect, but Send is async! Wait, or use SendNow.
+		time.Sleep(50 * time.Millisecond)
 		c.Banish()
```
Comment (chappjc) [inline server/comms/comms_test.go:334]: @buck54321 Here's a use case for SendNow.  I hit this bug while testing the send/outHandler refactoring commit, so I decided to put this in a separate commit.

File: dex/ws/wslink.go [hunk 6]
```diff
@@ -101,20 +123,47 @@ func (c *WSLink) Send(msg *msgjson.Message) error {
...
-		}
-	}()
+	// NOTE: without the stopped chan or access to the Context we are now racing
+	// after the c.Off check above.
+
+	select {
+	case c.outChan <- &sendData{b, writeErr}:
+	case <-c.stopped:
+		return ErrPeerDisconnected
+	}
```
Comment (chappjc) [inline dex/ws/wslink.go:133]: @buck54321 I squashed in a refactor of send and outHandler with the previous SendNow commit... breaking my own rules about squashing during review, but the SendNow thing really doesn't belong in this PR anyway.  We can take this whole commit out into another PR if that would be better for review.

File: server/asset/btc/utxo.go [hunk 7]
```diff
@@ -150,49 +154,66 @@ func (input *Input) spendsCoin(coinID []byte) (bool, error) {
...
+
+// A UTXO is information regarding an unspent transaction output.
+type UTXO struct {
+	*Output
+}
+
+// Confirmations returns the number of confirmations on this output's
+// transaction. See also (*Output).Confirmations. This function differs from the
+// Output method in that it is necessary to relocate the utxo after a reorg, it
+// may error if the output is spent.
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/btc/utxo.go:216]: Yeah, that seemed problematic, which is why I documented it.  I didn't want to make too many sweeping changes to asset, namely changing behavior of the previous methods so I didn't actually remove utxo.

File: server/asset/common.go [hunk 8]
```diff
@@ -75,21 +90,23 @@ type FundingCoin interface {
...
-	// Value is the contract value.
-	Value() uint64
-	// Address retrieves the receiving address of the contract.
-	Address() string
-	// FeeRate returns the transaction fee rate, in atoms/byte equivalent.
-	FeeRate() uint64
-	// Script is the contract redeem script.
-	Script() []byte
+	// RefundAddres is the refund address of the swap contract.
+	RefundAddress() string
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/common.go:105]: It's just guaranteed to be part of the contract, so the method makes sense.  The concrete types could implement it and we could remove it from the interface though.

File: server/asset/dcr/dcr.go [hunk 9]
```diff
@@ -87,7 +87,7 @@ type Backend struct {
 	// The backend provides block notification channels through it BlockChannel
 	// method. signalMtx locks the blockChans array.
 	signalMtx  sync.RWMutex
-	blockChans []chan *asset.BlockUpdate
+	blockChans map[chan *asset.BlockUpdate]struct{}
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/dcr/dcr.go:90]: Yeah, if it looks reasonable.  I'll circle back on this.

File: server/swap/swap_test.go [hunk 10]
```diff
@@ -1256,8 +1410,12 @@ func TestTxWaiters(t *testing.T) {
...
 	// Get the MatchNotifications that the swapper sent to the clients and check
 	// the match notification length, content, IDs, etc.
-	ensureNilErr(rig.ackMatch_maker(true))
-	ensureNilErr(rig.ackMatch_taker(true))
+	if err := rig.ackMatch_maker(true); err != nil {
+		t.Fatal(err)
+	}
+	if err := rig.ackMatch_taker(true); err != nil {
+		t.Fatal(err)
+	}
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap_test.go:1418]: I got frustrated by `ensureNilErr` among other helpers because the error/fatal line number was in side the helper function rather than the line where the helper was called, making debugging the test quite difficult.

**Comment by chappjc** [comment]
Thanks for the fast turnaround with the review.  I made some quick replies, but will get into the other comments shortly.

> I'd like to work towards a fully DB-based solution eventually. I understand that it will be a lot more work, but what do you think are the major barriers for that? Is it mostly just a matter of grinding out the new tables, methods, and data structures, or is there more to it?

In my initial planning for this, a DB based solution was shaping up to be considerably more complex and error prone, with additional data needed only for state restore rather than normal operation.  Reloading the matches would be fairly straightforward although still requiring new columns like swap confirm time for the `swapStatus` structures. For the most part, I think you could figure out which acks are still pending and resend those, but I see a lot of gotchas trying to make those inferences, and a ton of DB work to identify which matches need what.

On the other hand, restarting active coin waiters would prove to be quite difficult as that requires the client initiated message (either Init or Redeem) to queue up processInit or processRedeem in the fresh latency queue.  We could just wait for any live coin waiters before shutting down, but a minute (`txWaitExpiration`) is longer than I'd like for shutdown. 

Similarly, the completion/failure tracking data in `orderMatches` is just a single line in restore (`s.orders.orderMatches = state.OrderMatches`) that would be quite hairy otherwise and I'm not even certain we could regenerate it correctly from the DB.

Besides these challenges, it would be certain to be quite slow by comparison too since the tables will be large.

**Comment by chappjc** [comment]
oops, wrong button

File: server/swap/state.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+type BackupFile struct {
+	Name  string
+	Stamp int64
+}
+
+// LatestStateFile attempts to locate the swap state file with the most recent
+// time stamp (in the file name) from the given directory. The *BackupFile
+// return will be nil if the directory was read successfully but no matching
+// files were found.
+func LatestStateFile(dir string) (*BackupFile, error) {
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/state.go:46]: Mostly for historical and debugging reasons.

File: server/auth/auth.go [hunk 12]
```diff
@@ -470,25 +651,47 @@ func (auth *AuthManager) handleConnect(conn comms.Link, msg *msgjson.Message) *m
...
-					cancelRatio, auth.cancelThresh),
-			}
+			// return &msgjson.Error{
+			// 	Code: msgjson.AccountClosedError,
+			// 	Message: fmt.Sprintf("cancellation ratio (%f) exceeds threshold %f",
+			// 		cancelRatio, auth.cancelThresh),
+			// }
 		}
 	} else {
 		client.conn = conn
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/auth/auth.go:670]: Hmm, yes it seems that the `client.conn` field could be written concurrent with a read in `request` or `send` *if* the client was already in the `auth.conns` maps.  Will address that.

File: server/auth/auth.go [hunk 13]
```diff
@@ -470,25 +651,47 @@ func (auth *AuthManager) handleConnect(conn comms.Link, msg *msgjson.Message) *m
...
+	log.Debugf("User %s connected from %s with %d pending requests.",
+		acctInfo.ID, conn.IP(), len(pendingReqs))
+
+	// Send pending requests for this user.
+	for _, pr := range pendingReqs {
+		// Use the AuthManager method to send so that failed requests, which
+		// result in client removal, will go back into the client's pending
+		// requests. Subsequent requests that follow removal also fail and go
+		// back into the pending requests.
+		connectTimeout := 2 * time.Minute
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/auth/auth.go:683]: Left over from fiddling.  I think it should just use `DefaultConnectTimeout` here too.

File: server/swap/swap.go [hunk 14]
```diff
@@ -288,130 +357,491 @@ type Swapper struct {
...
+			ss.swapTime = encode.UnixTimeMilli(ssd.SwapTime)
+		}
+
+		if ssd.SwapConfirmTime != 0 {
+			ss.swapConfirmed = encode.UnixTimeMilli(ssd.SwapConfirmTime)
+		}
+
+		if redeemCoin := ssd.RedeemCoinIn; len(redeemCoin) > 0 {
+			assetID := ssd.RedeemAsset
+			redeem, err := s.coins[assetID].Backend.Input(redeemCoin)
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:435]: The idea was to skip spendsCoin check (note the `contractID` arg in `Redemption(redemptionID, contractID []byte)`), but that's not really an issue since `contractID` will necessarily be available of the redeem coin is.  Will try this route.  I think I had a different set of considerations in an earlier draft, so this is probably fine.

File: server/swap/swap.go [hunk 15]
```diff
@@ -260,14 +255,88 @@ func (s *orderSwapTracker) canceled(ord order.Order) {
...
+// Data input to handleInit and handleRedeem, which both start coin waiters with
+// latencyQ. Store handlerArgs before starting the waiter, and remove them when
+// the waiter completes or expires. On Swapper shutdown, the stored handlerArgs
+// should be saved in the state data after stopping the latencyQ.
+type handlerArgs struct {
+	User account.AccountID
+	Msg  *msgjson.Message // init or redeem inferred from Msg.
+}
+
+// client generated message IDs are not globally unique.
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:312]: I didn't think this was trouble as the goal of this was just to key uniquely.

But since you mention reconnects, I would expect client-originating requests would have handlers registered by ID that would not be be lost on disconnect (only dexc *shutdown*).  If this association is lost in the client, the client might not know what to do with any responses from `processInit` and `processRedeem` (whether `respondError` or `respondSuccess`).  Off hand I'm don't know what happens to the client's request handler map when the connection goes down, but I thought it was fine as long as the software didn't shutdown.

But otherwise, `processInit` and `processRedeem` don't need the client's msg ID provided by the original `handleInit`/`handleRedeem` to do their jobs, which consist of validating the contract/redeem via the asset backends, storing the validated data, and sending an ack request to the counter party with a fresh server-generated msg ID.  The only possible issue is as I mentioned with the success/fail message sent to the user having an ID that the client no longer recognizes, or even associates with a different request.

With the server-originating requests, I regenerate the msg ID when firing up the live ackers, but it's clearly more of an issue for the client-originating requests.  Do you think this can be handled client side?

[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:312]: Yeah, the response handlers map persists across reconnects in client/comms, but of course this is lost on dexc shutdown or anything that makes a fresh `wsConn`.

[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:312]: Looking at client/core, it seems that if these coin waiters are live on the server, then either `(*trackedTrade).swapMatches` or `(*trackedTrade).redeemMatches` are sitting at `signAndRequest` waiting to receive and unmarshal a `msgjson.Acknowledgement`.  And it seems `signAndRequest` would not be erroring as long as the request `Send` completed without err (and it would have if the server has has a coin waiter running from it).  The client just gives the connection `time.Minute*5` according to `client/comms.(*wsConn).logReq` to respond.

File: server/asset/btc/btc.go [hunk 16]
```diff
@@ -403,35 +456,117 @@ func (btc *Backend) utxo(txHash *chainhash.Hash, vout uint32, redeemScript []byt
...
-	}, nil
+	}
+	return &UTXO{out}, nil
 }
 
-// input gets the transaction input.
-func (btc *Backend) input(txHash *chainhash.Hash, vin uint32) (*Input, error) {
+// newTXIO creates a TXIO for a transaction, spent or unspent. The caller must
+// set the maturity field.
+func (btc *Backend) newTXIO(txHash *chainhash.Hash) (*TXIO, int64, error) {
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/btc/btc.go:465]: Actually, maturity depends on the output not just the transaction.  `maturity` needs to be moved out of `TXIO` and into the containing type like `Output`.  Ticket change outputs (`sstxchange` tagged) are spendable with 1 confirm (`SStxChangeMaturity`) while the actual ticket output (`stakesubmission` tagged) are spendable in `TicketMaturity`, plus those `stakesubmission` outputs can only be spent by a revocation or vote and thus should be considered unspendable for DEX purposes.  `CoinbaseMaturity` only refers to outputs of coinbase and stakebase/votes (`stakegen` outputs).

I think this needs to be addressed carefully.  I suggest we fix this in a follow up PR.

[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/btc/btc.go:465]: Ah, for btc, addressed.

File: server/swap/swap.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:404]: Yeah, Redemption works fine even if the spendsCoin check probably is not needed in swap state restore.  I'm pretty sure I created the Input method when I was pursing some (fantastically complicated) GobDecoders.... turned out that it was just far easier to copy a bit of extra data.  Anyway, `Input`, `Output` and `UnspendContract` methods of the `asset.Backend` interface are now removed.

File: server/auth/auth.go [hunk 17]
```diff
@@ -289,52 +405,112 @@ func (auth *AuthManager) Send(user account.AccountID, msg *msgjson.Message) {
...
+	}
+
+	auth.storage.CloseAccount(user /*client.acct.ID*/, rule)
+
+	// We do NOT want to do disconnect if the user has active swaps.  However,
+	// we do not want the user to initiate a swap or place a new order, so there
+	// should be appropriate checks on order submission and match/swap
+	// initiation (TODO).
+
+	// TODO: notify client of penalty / account status change?
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/auth/auth.go:514]: Opened issue https://github.com/decred/dcrdex/issues/444

File: server/asset/btc/utxo.go [hunk 18]
```diff
@@ -150,49 +154,66 @@ func (input *Input) spendsCoin(coinID []byte) (bool, error) {
 	return txIn.prevTx == *txHash && txIn.vout == vout, nil
 }
 
-// A UTXO is information regarding an unspent transaction output.
-type UTXO struct {
+// Output represents a transaction output.
+type Output struct {
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/btc/utxo.go:158]: The assets are now cleaned up quite a bit but there are a couple TODOs for another PR:

- address maturity being output specific for decred stake transactions
- remove the UTXO type completely

An update for the `btc.Backend.blockChans` field is incoming though.

File: server/swap/swap.go [hunk 19]
```diff
@@ -1466,31 +2070,38 @@ func (s *Swapper) revoke(match *matchTracker) {
 	s.unlockOrderCoins(match.Taker)
 	s.unlockOrderCoins(match.Maker)
 
+	sendRevReq := func(user account.AccountID, req *msgjson.Message, params *msgjson.RevokeMatch) {
+		ack := &messageAcker{
+			user:    user,
+			match:   match,
+			params:  params,
+			isMaker: false,
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:2078]: Whoops.

File: server/swap/swap.go [hunk 20]
```diff
@@ -963,6 +1426,11 @@ type messageAcker struct {
 // signature and updating the (order.Match).Sigs record. This is required by
 // processInit, processRedeem, and revoke.
 func (s *Swapper) processAck(msg *msgjson.Message, acker *messageAcker) {
+	// Remove the live acker from Swapper's tracking, but consider keeping it if
+	// there is any sense to the client retrying the response if this function
+	// responds with an error.
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:1431]: I'm just removing the "consider" parts.  It's just rambling.

File: server/swap/swap.go [hunk 21]
```diff
@@ -501,7 +958,8 @@ func (s *Swapper) processBlock(block *blockNotification) {
 		// Remove the completed match. Note that checkInaction may also remove
 		// matches, so this entire function must lock even if there are no
 		// completions.
-		delete(s.matches, match.ID().String())
+		log.Debug("Deleting completed match", match.ID())
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:961]: `Debug` will just print each argument delimited by a space so it's actually what I indented, but I'll switch to `log.Debugf("Deleting completed match %v", match.ID())` so this doesn't stand out.

File: server/swap/swap.go [hunk 22]
```diff
@@ -277,141 +277,598 @@ type Swapper struct {
...
+			auth.DefaultConnectTimeout, expireFunc)
+		// NOTE: The user must reconnected to send the ack request. processAck
+		// and processMatchAcks themselves won't require connectivity since
+		// there should be no error response (already validated by virtue of
+		// being in the state's live ackers) and they just save the ack
+		// signatures.
+	}
+
+	// Live coin waiters started by the comms handlers for client init and
+	// redeem messages
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:548]: This was supposed to be like a section heading, although a long one.

File: server/swap/swap.go [hunk 23]
```diff
@@ -277,141 +277,598 @@ type Swapper struct {
...
+		LiveAckers:    s.liveAckers,
+		LiveWaiters:   s.liveWaiters,
+	}
+
+	enc := gob.NewEncoder(f)
+	if err = enc.Encode(st); err != nil {
+		log.Errorf("Failed to save swap state to file %v: %v", fName, err)
+		return
+	}
+	f.Close()
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:630]: Thanks for the link.  Although this `f.Close` was stray and apparently not harmful, I like the "Update 2" suggestion at that page to run `f.Sync` after writing but before return, and keep the `defer f.Close()`.

File: server/swap/swap.go [hunk 24]
```diff
@@ -277,141 +277,598 @@ type Swapper struct {
...
+func NewSwapper(cfg *Config) (*Swapper, error) {
+	// Verify the directory where swap state will be saved.
+	inf, err := os.Stat(cfg.DataDir)
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("data folder %q does not exist", cfg.DataDir)
+	}
+	if !inf.IsDir() {
+		return nil, fmt.Errorf("path %q is not a directory", cfg.DataDir)
+	}
+
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap.go:345]: I started down this rabbit hole and it gets hairy and/or os specific very quickly.  I decided it wan't really worth it, although a quick and dirty check would be to touch a new file in the folder.  Open to suggestions.

File: server/swap/state.go [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+	// Stamp is the millisecond unix epoch time stamp of the state.
+	Stamp int64
+}
+
+// LatestStateFile attempts to locate the swap state file with the most recent
+// time stamp (in the file name) from the given directory. The *BackupFile
+// return will be nil if the directory was read successfully but no matching
+// files were found.
+func LatestStateFile(dir string) (*BackupFile, error) {
+	cwd, err := os.Open(dir)
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/state.go:56]: Interesting question!  It seems that there is a distinct close path in os/file_unix.go for directories, but [ends up as a noop](https://golang.org/src/os/dir_unix.go#L27).  But in darwin it does somehting.  Anyway, good catch.

File: server/cmd/dcrdex/main.go [hunk 26]
```diff
@@ -94,8 +96,58 @@ func mainCore(ctx context.Context) error {
...
+		// 	return fmt.Errorf("aborted input: %v", err)
+		// }
+
+		var doLoad bool
+		switch strings.ToLower(promptResp) {
+		case "y", "yes":
+			doLoad = true
+		case "n", "no":
+		case "":
+			return fmt.Errorf("input aborted")
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/cmd/dcrdex/main.go:129]: Yeah, but there's no `fmt.Error`.  Could use `errors.New` I suppose although we're not importing `errors` presently...  Why not, will just use errors.New

File: server/auth/auth.go [hunk 27]
```diff
@@ -470,25 +786,56 @@ func (auth *AuthManager) handleConnect(conn comms.Link, msg *msgjson.Message) *m
...
+		acctInfo.ID, conn.IP(), len(pendingReqs), len(pendingMsgs))
+
+	// Send pending requests for this user.
+	for _, pr := range pendingReqs {
+		// Use the AuthManager method to send so that failed requests, which
+		// result in client removal, will go back into the client's pending
+		// requests. Subsequent requests that follow removal also fail and go
+		// back into the pending requests.
+		connectTimeout := DefaultConnectTimeout
+		// Decrement the connect timeout for repeated attempts.
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/auth/auth.go:816]: Yup, you got it.  I'll clarify the "Each connection reduces connectTimeout" sentence.

File: server/swap/swap_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply server/swap/swap_test.go:1410]: I recently found the method off of testing `t.Helper()` which may allow helpers to be used while not messing up the line numbers.
https://golang.org/pkg/testing/#B.Helper

[the remark this answers is not in the document]
Comment (chappjc) [reply server/swap/swap_test.go:1410]: Wow, nice find @JoeGruffins

File: dex/ws/wslink.go [hunk 28]
```diff
@@ -298,12 +289,39 @@ func (c *WSLink) outHandler(ctx context.Context) {
...
+			case sd := <-c.outChan:
+				outQueue = append(outQueue, sd)
+			default:
+				break out
+			}
+		}
+		// Attempt sending all queued outgoing messages.
 		for _, sd := range outQueue {
-			relayError(sd.ret, ErrPeerDisconnected)
+			write(sd)
```
[the remark this answers is not in the document]
Comment (buck54321) [reply dex/ws/wslink.go:324]: You are preserving the map now, but not in all cases. If a send error is encountered server-side before the client has reconnected, `removeClient` would already have been called, and we lose the `respHandlers`. Should we open an issue for this?

[the remark this answers is not in the document]
Comment (chappjc) [reply dex/ws/wslink.go:324]: Is this comment meant to be placed on the diff in https://github.com/decred/dcrdex/pull/406/commits/0ddc52f5dc6f6eb49b46db41a2da7c8e0f8b92dd?  I'm trying to discern if there's a connection between the two spots.

Although I do see that between `auth.connMtx.Unlock()` in the branch of `(*AuthManager).handleConnect` and the `auth.addClient` call to replace the clientInfo with the fresh connection that the map could be modified because it's a shallow copy.  Is that what you are referring to here?  I can make a deep copy of the map instead to resolve this.
